funktionale-try/src/test/kotlin/org/funktionale/tries/TryTest.kt

```diff
@@ -54,14 +54,75 @@ class TryTest {
         assertEquals(failure.orElse { Success(5) }.get(), 5)
     }
 
-    @Test fun foreach() {
-        success.foreach { assertEquals(10, it) }
-        failure.foreach { fail() }
+    @Test fun foreach_success_sideEffect() {
+        var wasInside = false
+        success.foreach { wasInside = true }
+        assertTrue(wasInside)
     }
 
-    @Test fun onEach() {
-        assertEquals(success.onEach { assertEquals(10, it) }.get(), 10)
-        assertTrue(failure.onEach { fail() }.isFailure())
+    @Test fun foreach_failure_sideEffect() {
+        var wasInside = false
+        failure.foreach { wasInside = true }
+        assertFalse(wasInside)
+    }
+
+    @Test fun foreach_success_throwInside() {
+        try {
+            success.foreach { throw RuntimeException("thrown inside") }
+        } catch (e: Exception) {
+            assertTrue(e is RuntimeException)
+            assertEquals(e.message, "thrown inside")
+        }
+    }
+
+    @Test fun foreach_failure_throwInside() {
+        try {
+            failure.foreach { throw RuntimeException("thrown inside") }
+        } catch (e: Exception) {
+            assertTrue(e is NumberFormatException)
+        }
+    }
+
+    @Test fun onEach_success_sideEffect() {
+        var wasInside = false
+        success.onEach { wasInside = true }
+        assertTrue(wasInside)
+    }
+
+    @Test fun onEach_failure_sideEffect() {
+        var wasInside = false
+        failure.onEach { wasInside = true }
+        assertFalse(wasInside)
+    }
+
+    @Test fun onEach_success_throwInside() {
+        try {
+            success.onEach { throw RuntimeException("thrown inside") }.get()
+        } catch (e: Exception) {
+            assertTrue(e is RuntimeException)
+            assertEquals(e.message, "thrown inside")
+        }
+    }
+
+    @Test fun onEach_failure_throwInside() {
+        try {
```

Comment:
Overall, the tests look good. :+1:

There's one thing to consider: try-catch works sort of like a conditional. If the line `failure.onEach { throw RuntimeException("thrown inside") }.get()` never throws an exception, `assertTrue(e is NumberFormatException)` is not executed. What should be done instead is that the test verifies first that there IS an exception being thrown.

You can leverage [try as an expression](https://kotlinlang.org/docs/reference/exceptions.html#try-is-an-expression) or [funKTionale's Try](https://github.com/MarioAriasC/funKTionale/blob/master/funktionale-try/src/main/kotlin/org/funktionale/tries/Try.kt) for first capturing the exception, and then doing the assertions outside of the `catch` block. There's also [TestNG excepted exceptions](http://www.mkyong.com/unittest/testng-tutorial-2-expected-exception-test/), which might help.

Example format:

```kotlin
val ex: Exception? = try {
  doStuff()
  null
} catch (e: Exception) {
  e
}
assertEquals(ex!!.message, "my message")
```